---
title: "R Notebook"
output:
  html_document:
    error: no
    message: no
    toc: yes
    toc_depth: 4
    toc_float: yes
    warning: no
  pdf_document:
    toc: yes
    toc_depth: '4'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```

This Rnotebooks cover the later part of 3' motif finding project, it contains description about:

* **motif search** by MEME 

* **motif occurence observation** by MAST

* **motif comparisons** by TOMTOM

Three steps above use tools from MEME-suite, which is a basic and commonly used toolset for motif discovery operation. Shell scripts for running those tools will be described here, but they can't run properly in Rstudio. The user has to run the scripts in linux-shell.

**main reasons why this toolset was chosen** 

* user-friendly

* simple commands, 

* detailed description and example 

* all-in-one toolset

* was used by the original paper

**Note:** Alternative motif operation tools other than MEME-suite has not been tested yet.

The later part of this notebook describes post-processing pipeline which will 

* **gather and rearrange MEME results** into simplified and tidy tables. 

###MOTIF FINDING

After we separate 500bp upstream and downstream sequences into clusterà¸«, we then search for the motif sequences in each cluster. MEME will look for the short sequences which appear in every sequence in the cluster aka motif and report them into MEME-html/xml/txt formats. These files will be used as inputs for downstream process such as counting motif occurence in region other than the discovery cluster or comparing motif to database. 

First, we will start with building background base frequency from whole genome sequences of 500bp upstream and downstream. Then feed the clustered sequence files to MEME to find the motifs.

**Starting File**
* background file: background_down.fasta/background_up.fasta 
* clustered sequence cluster: cluster_1025_Ward (extract to the home directory)

###BACKGROUND BUILDING
This function will convert 500bp sequence set downloaded from ensmbl database to FASTA file for background frequncy calculation. (Actually it can also be used to convert any gene-sequence fetched from ensmbl to fasta)

define function
```{r}
create_background_from_sequence_table<-function(sequence_file){
  #print out FASTA file      
      output_tab<-sequence_file%>%pull(coding_gene_flank)%>%DNAStringSet()
      names(output_tab)<-sequence_file$ensembl_gene_id
      region_vector<-unlist(strsplit(deparse(substitute(sequence_file)),"_"))[1]
      tab_name<-paste0("background_",region_vector,".fasta")
      writeXStringSet(output_tab, tab_name, format="fasta")
      
}
```

```{r}
create_background_from_sequence_table(down_500)
create_background_from_sequence_table(up_500)
```

####Background model
background model calculation

This step was done by using MEME suite command fasta-get-markov to create background model. -m argument control markov model order. These are shell command so they need to be typed outside Rstudio. 
```{bash eval=FALSE}
fasta-get-markov -m 1 < background_down.fasta > bg_down-model-1
fasta-get-markov -m 1 < background_up.fasta > bg_up-model-1

```

#Background_motif script
This is the script for motif discovery for background file. The resuts are used to compare the effect of inputs (whole genome [background] vs clusters). The speculation is that very common motifs could be found from the whole genome set, any noise should have no effect to them.

**background_meme.sh**
to run: bash background_meme.sh
```{bash}
#!/bin/sh
#set working dir as the current dir
home_dir=$(pwd)
echo begin_down
date
mkdir background_down
#-dna: specify data in fasta file
#-revcomp: consider reverse complement
#-mod anr: Any Number of Repetitions (a motif might repeats multiple times in one sequence)
#-evt: e-value treashold
#-p 20: paralellise the task to speed things up
#bfile:  background file
#maxw, minw: maximum and minimum width
/usr/local/meme5/bin/meme ./background_down.fasta -dna -revcomp -mod anr -oc $home_dir/background_down -nmotifs 50 -evt 0.005 -brief 1000000 -p 20 -bfile $home_dir/bg_down-model-1 -maxw 10 -minw 6
echo begin_up
date
mkdir background_up
/usr/local/meme5/bin/meme ./background_up.fasta -dna -revcomp -mod anr -oc $home_dir/background_up -nmotifs 50 -evt 0.005 -brief 1000000 -p 20 -bfile $home_dir/bg_down-model-1 -maxw 10 -minw 6
echo finish at:
date

```


###SCRIPT

```{bash}
#!/bin/sh
echo begin_down
date
#choose directory
dir_list=$(ls -d cluster_20/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do 
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd
#loop for each cluster resulted from clustering
for file in $home_dir/$dir/*down.fasta; 
do 
echo ${file}
echo start processing $(basename "$file") at:
date
#-dna: specify data in fasta file
#-revcomp: consider reverse complement
#-mod anr
#-p 20: paralellise the task to speed things up
/usr/local/meme5/bin/meme ${file} -dna -revcomp -mod anr -oc $(basename "$file" .fasta) -nmotifs 30 -evt 0.005 -brief 1000000 -p 20 -bfile $home_dir/bg_down-model-1 -maxw 10 -minw 6
echo finish at:
date
done 
cd ".."
done

echo begin_up
date
#choose directory
dir_list=$(ls -d cluster_20/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd
#loop for each cluster resulted from clustering
for file in $home_dir/$dir/*up.fasta;
do
echo ${file}
echo start processing $(basename "$file") at:
date
#-dna: specify data in fasta file
#-revcomp: consider reverse complement
#-mod anr
#-p 20: paralellise the task to speed things up
/usr/local/meme5/bin/meme ${file} -dna -revcomp -mod anr -oc $(basename "$file" .fasta) -nmotifs 30 -evt 0.005 -brief 1000000 -p 20 -bfile $home_dir/bg_up-model-1 -maxw 10 -minw 6
echo finish at:
date
done
cd ".."
done

```

###MEME

```{bash}
vi meme.sh
chmod +x meme.sh
nohup ./meme.sh &
```

##MOTIF SEARCH

MAST SCRIPT
```{bash}
#!/bin/sh
echo begin
date
#choose directory
dir_list=$(ls -d cluster_20/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd

#loop for each cluster resulted from clustering
for query in $(find $home_dir/$dir -name '*meme.html' |sort);
do
echo $query
target01=$(find $home_dir/$dir -name '*up.fasta'| sort)
for target in $target01
do
echo target file $target
rm -r $(dirname $query)/mast_out/
mkdir $(dirname $query)/mast_out/
/usr/local/meme5/bin/mast ${query} ${target} -oc $(dirname $query)/mast_out/$(basename $target .fasta) -bfile $home_dir/bg_up-model-1
done
target02=$(find $home_dir/$dir -name '*down.fasta'| sort)
for target in $target02
do
echo target file $target
rm -r $(dirname $query)/mast_out/
mkdir $(dirname $query)/mast_out/
/usr/local/meme5/bin/mast ${query} ${target} -oc $(dirname $query)/mast_out/$(basename $target .fasta) -bfile $home_dir/bg_down-model-1
done
done
cd ".."
done

```

###MAST

```{bash}
#!/bin/sh
echo begin
date
#choose directory
dir_list=$(ls -d cluster_19/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd

#loop for each cluster resulted from clustering
for query in $(find $home_dir/$dir -name '*meme.html' |sort);
do
echo $query
target01=$(find $home_dir/$dir -name '*up.fasta'| sort)
for target in $target01
do
echo target file $target
#rm -r $(dirname $query)/mast_out/
#mkdir $(dirname $query)/mast_out/
/usr/local/meme5/bin/mast ${query} ${target} -oc $(dirname $query)/mast_out/$(basename $target .fasta) -bfile $home_dir/bg_up-model-1
done
target02=$(find $home_dir/$dir -name '*down.fasta'| sort)
for target in $target02
do
echo target file $target
#rm -r $(dirname $query)/mast_out/
#mkdir $(dirname $query)/mast_out/
/usr/local/meme5/bin/mast ${query} ${target} -oc $(dirname $query)/mast_out/$(basename $target .fasta) -bfile $home_dir/bg_down-model-1
done
done
cd ".."
done

```

meme2meme
```{r}
#!/bin/sh
echo begin
date
#choose directory
dir_list=$(ls -d cluster_19/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd
echo merge upstream_motifs
meme_up=$(find $home_dir/$dir -path '*up/meme.html'| sort)
echo $meme_list
/usr/local/meme5/libexec/meme-5.0.0//meme2meme $meme_up -bg $home_dir/bg_up-model-1 > meme_up.txt
echo merge downstream_motifs
meme_down=$(find $home_dir/$dir -path '*down/meme.html'| sort)
echo $meme_list
/usr/local/meme5/libexec/meme-5.0.0//meme2meme $meme_down -bg $home_dir/bg_down-model-1 > meme_down.txt
done
```

tomtom
```{r}
#!/bin/sh
echo begin
date
#choose directory
dir_list=$(ls -d cluster_19/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd
mkdir $home_dir/$dir/tomtom_out
echo self_comparison
mkdir $home_dir/$dir/tomtom_out/down_down/
/usr/local/meme5/bin/tomtom meme_down.txt meme_down.txt -oc $home_dir/$dir/tomtom_out/down_down/
mkdir $home_dir/$dir/tomtom_out/up_up/
/usr/local/meme5/bin/tomtom meme_up.txt meme_up.txt -oc $home_dir/$dir/tomtom_out/up_up/
echo pair_comprison
mkdir $home_dir/$dir/tomtom_out/up_down/
/usr/local/meme5/bin/tomtom meme_up.txt meme_down.txt -oc $home_dir/$dir/tomtom_out/up_down/
mkdir $home_dir/$dir/tomtom_out/down_up/
/usr/local/meme5/bin/tomtom meme_down.txt meme_up.txt -oc $home_dir/$dir/tomtom_out/down_up/
done
```

##TABLE OF MOTIFS X GENE EXPRESSION
Generate mean expression table
need: cluster profile
expression table
```{r}
WardD2.cluster_table
```
gene name with cluster number
```{r}
cluster_number<-as.tibble(WardD2.cluster_table[,"cluster_19"])
cluster_number$rowname<-rownames(WardD2.cluster_table)
```
gene name with expression
```{r}
expression_table<-as.tibble(rownames_to_column(as.data.frame(init_table_142.mat)))
```

leftjoin
```{r}
expression_table_with_cluster_number<-left_join(cluster_number,expression_table, by="rowname")
```

```{r}
write.table(expression_table_with_cluster_number, file=paste0("./expression_table_with_cluster_number_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t")
```

set rownames
```{r}
rownames(expression_table_with_cluster_number)<-expression_table_with_cluster_number$rowname
```

```{r}
summarized_expression_table<-expression_table_with_cluster_number%>%select(-rowname) %>% group_by(value) %>% summarise_all(funs(mean(., na.rm = TRUE)))
```

```{r}
write.table(summarized_expression_table, file=paste0("./summarized_expression_table_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t")
```

```{r}
summarized_expression_table
```


###create motifs table

####methods
read XML results files and extract their information into small tables, then join the tables into one big table.


```{r message=FALSE, warning=FALSE, paged.print=FALSE}
library(XML)
library(methods)
```

####prepare iteration
```{r}
clust_all_subfolder_list<-list.dirs(path = ".", full.names = FALSE, recursive = TRUE)
clust_all_subfolder_list<-grep("in",clust_all_subfolder_list,value = TRUE)
clust_all_subfolder_list<-grep("mast",clust_all_subfolder_list,value = TRUE,invert = TRUE)
clust_all_subfolder_list<-grep("cluster_19_005",clust_all_subfolder_list,value = TRUE)
all_xml_list<-list.files(path = ".", pattern = "meme.xml", all.files = TRUE, full.names = TRUE, recursive = TRUE, include.dirs = FALSE, no.. = TRUE)
all_xml_list<-grep("cluster_19_005",all_xml_list,value = TRUE)

clust_all_subfolder_list
all_xml_list

```
####inputs
meme .xml results files
```{r warning=FALSE}
#set_dir<-"./cluster_0626/"
#setwd("./cluster_0626/")

#setting up iteration
clust_max_list<-list.dirs(path = ".", full.names = FALSE, recursive = FALSE)

clust_all_subfolder_list<-list.dirs(path = ".", full.names = FALSE, recursive = TRUE)
clust_all_subfolder_list<-grep("in",clust_all_subfolder_list,value = TRUE)
clust_all_subfolder_list<-grep("mast",clust_all_subfolder_list,value = TRUE,invert = TRUE)
clust_all_subfolder_list<-grep("cluster_19_005",clust_all_subfolder_list,value = TRUE)
all_xml_list<-list.files(path = ".", pattern = "meme.xml", all.files = TRUE, full.names = TRUE, recursive = TRUE, include.dirs = FALSE, no.. = TRUE)
all_xml_list<-grep("cluster_19_005",all_xml_list,value = TRUE)
clust_all_subfolder_list
all_xml_list

#data gethering loop
master_table<-NA
for (i in clust_max_list){
  folname<-i
  subfolname<-grep(paste0("^",i), clust_all_subfolder_list,value=TRUE)
  table.tmp<-NA
  table_part.tmp<-NA
  for (i in subfolname){
    clustname.temp<-gsub(i,pattern = "cluster_[0-9]/",replacement = "")
    cluster.main=unlist(strsplit(clustname.temp,"/"))[1]
    cluster.sub<-unlist(strsplit(clustname.temp,"/"))[2]
    subcluster<-paste(unlist(strsplit(cluster.sub,"_"))[1:4],collapse = "_")
    region<-paste(unlist(strsplit(cluster.sub,"_"))[5])
    
    xml_file<-grep(clustname.temp, all_xml_list,value=TRUE)
    xml.tmp<-xmlParse(xml_file)
    print(xml_file)
    motif_set<-getNodeSet(xml.tmp,"//motif")
    motif_desc<-t(sapply(motif_set, xmlAttrs))
    
    motif_desc<-motif_desc[,c("name","width","sites","e_value")]
    reg_set<-(getNodeSet(xml.tmp,"//regular_expression"))
    reg_extracted<-(sapply(reg_set, xmlValue, trim = TRUE))
    
    train_set<-(getNodeSet(xml.tmp,"//training_set"))
    seq_filename<-(sapply(train_set, xmlAttrs))[1]
    
    RootNode <- xmlRoot(xml.tmp)

    ChildNodes <- xmlChildren(RootNode)
    
    seqlist.ChildNodes<-xmlChildren(ChildNodes[["training_set"]])
    genid.tmp<-unlist(seqlist.ChildNodes)[-c(1,length(unlist(seqlist.ChildNodes)))]
    nameidlist<-data.frame(t(sapply(genid.tmp,xmlAttrs)))
    seqlist.ChildNodes<-xmlChildren(ChildNodes[[4]])
    seqmotiflist<-data.frame(t(sapply(seqlist.ChildNodes, xmlAttrs)))
    genetab<-left_join(nameidlist,seqmotiflist,by=c("id"="sequence_id"))[,c(2,5,6)]
    
    if(is.null(dim(motif_desc))){
      motif_desc<-t(motif_desc)
    }
    print("check1")
    table_part.tmp<-cbind(main_dir=set_dir,cluster=cluster.main,subcluster=cluster.sub,region=region,fasta_file=seq_filename,motif_desc, regular_expression= reg_extracted)
    print("check2")
   
    
    if (is.na(table.tmp)==TRUE){
      table.tmp<-table_part.tmp

    }
    else{
      table.tmp<-rbind(table.tmp,table_part.tmp)
    }
  }
    print("check3")
    if (is.na(table.tmp)){
      next
    }
    print("check4")
    if (is.na(master_table)){
      master_table<-table.tmp
    }
    else{
      master_table<-rbind.data.frame(master_table,table.tmp)%>% distinct()
    }

  write.table(master_table, file="./meme_table.txt", sep="\t", row.names = FALSE)
}


```


##Outcome
```{r}
head(master_table)
```


```{r}
master_table<-as.tibble(master_table)
```

```{r}
summarized_table<-master_table %>% group_by(subcluster) %>% summarize(motif_list=paste(sort(unique(name)), collapse = " | ")) %>% separate(subcluster, into=c("cluster","region"), sep="_in_10_") %>% spread(region,motif_list)
summarized_table
```

```{r}
write.table(master_table, file=paste0("./motif_table_full_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t", row.names = FALSE)


```
```{r}
write.table(summarized_table
, file=paste0("./motif_table_summarized_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t", row.names = FALSE, quote=F)
```


###Simplify tidy motif tables

```{r}
master_table<-read.table(file=paste0("./motif_table_full_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t", header = T)
master_table[1:10,]
```
```{r}
dim(master_table)
```


```{r}
cluster_info<-master_table[,c("subcluster","name","sites")]
```

```{r}
dim(cluster_info)
```

```{r}
write.table(cluster_info, file=paste0("./cluster_info_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t", quote=F, row.names = F)
```
```{r}
write.table(motif_info_1, file=paste0("./cluster_info_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t", quote=F, row.names = F)

```


```{r}
motif_info_1<-cluster_info%>%separate(subcluster,sep="_",into=c("cluster","value","2","3","direction"))%>%select(-c("cluster","2","3"))
```
```{r}
cluster_info_2<-master_table[,c("subcluster","name","e_value")]
motif_info_2<-cluster_info_2%>%separate(subcluster,sep="_",into=c("cluster","value","2","3","direction"))%>%select(-c("cluster","2","3"))
```

```{r}
write.table(motif_info_2, file=paste0("./cluster_info_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t", quote=F, row.names = F)

```


```{r}
dim(motif_info_2)
```

```{r}
head(motif_info_2)
```



```{r}
motif_and_expression<-left_join(motif_info_1,summarized_expression_table,by="value")
```
```{r}
write.table(motif_and_expression, file=paste0("./motif_and_expression_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t")
```

Redefining table
keep them super tidy
```{r}
mast_list<-list.files(path = ".", pattern = "mast.xml", all.files = TRUE, full.names = TRUE, recursive = TRUE, include.dirs = FALSE, no.. = TRUE)
```

```{r}
extract_xml_attribute<-function(x,y){
  x %>% map(xmlAttrs) %>% map_chr(y)
}

get_mast_table<-function(i){
xml.tmp<-xmlParse(file = i)
source<-i %>% str_split(pattern = "/") %>% reduce(c) %>% extract2(3)
target<-i %>% str_split(pattern = "/") %>% reduce(c) %>% extract2(5)
motif_set<-getNodeSet(xml.tmp,"//motif")
motif_id<-extract_xml_attribute(motif_set, 2)
test1<-read_xml(i) %>% xml_child(search="sequences")
  if(is_empty(xml_children(test1))==TRUE){
  test2<-NA
  } else {
    test2<-test1 %>% xml_children() %>% map(xml_children)%>%map(xml_children) %>% map(xml_attr, "idx") %>% reduce(c)
  }
idx<-read_xml(i) %>% xml_child(search="motifs") %>% xml_children() %>% length() %>% seq() 
idx<-as.character(idx-1)
count<-NA
for(i in seq_along(idx)){
  count[i]=length(which(na.omit(test2)==idx[i]))
}
motif_nsite<-cbind.data.frame(idx,count) 
temp_frame<-cbind.data.frame(source, motif_id,target, count=motif_nsite) 

}

```
```{r}
extract2  
```

####extraction loop

Running loop and return a nested list of small tables.

```{r message=FALSE, warning=FALSE}
testvec<-map(mast_list,get_mast_table)
```

```{r message=FALSE, warning=FALSE}
full_table<-testvec%>%reduce(bind_rows)
```

```{r}
full_table_tidy<-full_table%>%separate(source,sep="_",into=c("cluster","source","2","3","source_direction"))%>%select(-c("cluster","2","3"))%>%separate(target,sep="_",into=c("cluster","target","2","3","target_direction"))%>%select(-c("cluster","2","3"))
```

```{r}
full_table_tidy
```


```{r}
write.table(full_table_tidy, file=paste0("./tidy_motif_count",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t")
```

gene-motif-count
```{r}
extract_xml_attribute<-function(x,y){
  x %>% map(xmlAttrs) %>% map_chr(y)
}

get_mast_table_gene_level<-function(i){
xml.tmp<-xmlParse(file = i)
#print(i) #for tracking in debug process
source<-i %>% str_split(pattern = "/") %>% reduce(c) %>% extract2(3)
motif_set<-getNodeSet(xml.tmp,"//motif")
motif_id<-extract_xml_attribute(motif_set, 2)
idx<-read_xml(i) %>% xml_child(search="motifs") %>% xml_children() %>% length() %>% seq() 
idx<-as.character(idx-1)
motif_idx_set<-cbind.data.frame(idx,motif_id)

test1<-read_xml(i) %>% xml_child(search="sequences")
if(is_empty(xml_children(test1))==TRUE){
  temp_frame<-as.tibble(NA)
}
else{
idx_set<-test1 %>% xml_children() %>% map(xml_children)%>%map(xml_children) %>% map(xml_attr, "idx")%>%map_if(is_empty,~ NA_character_)
target_gene<-test1 %>% xml_children() %>% map(xml_attr, "name")
target_set<-map2(target_gene,idx_set,cbind.data.frame) %>% map(drop_na) %>% reduce(bind_rows)%>%rename("gene"=1)%>%rename("idx"=2)

temp_frame<-cbind.data.frame(source, motif_idx_set) %>% left_join(target_set, by="idx") %>% drop_na() %>% group_by(source,gene,motif_id)%>%summarise(idx=n())
}
}
```



```{r message=FALSE, warning=FALSE}
gene_level_mast_set<-map(mast_list,get_mast_table_gene_level)
```

```{r message=FALSE, warning=FALSE}
full_table_gene_level_tidy<-gene_level_mast_set%>%map(drop_na)%>%reduce(bind_rows)
```

```{r}
write.table(full_table_gene_level_tidy, file=paste0("./tidy_motif_count_gene_level",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t")
```


```{r}
full_table_gene_level_tidy
```

```{r}
expression_table_with_cluster_number
```


```{r}
full_table_gene_level_summarised<-full_table_gene_level_tidy%>%select(-value)%>%group_by(gene,motif_id)%>%summarise(idx=sum(idx))%>%rename(count=idx)
```

```{r}
write.table(full_table_gene_level_summarised, file=paste0("./summarised_motif_count_gene_level",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t")
```

join with expression table
```{r}
gene_level_motif_and_expression<-full_table_gene_level_tidy%>%select(-value)%>% left_join(expression_table_with_cluster_number,by=c("gene"="rowname"))%>%rename(count=idx)%>%rename(cluster=value)
```



```{r}
write.table(gene_level_motif_and_expression, file=paste0("./gene_level_motif_and_expression",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t")
```

```{r}
summarised_gene_level_motif_and_expression<-full_table_gene_level_summarised%>% left_join(expression_table_with_cluster_number,by=c("gene"="rowname"))%>%rename(cluster=value)
```

```{r}
write.table(summarised_gene_level_motif_and_expression, file=paste0("./summarised_gene_level_motif_and_expression_",unlist(str_split(getwd(),pattern="/"))[4],".txt"), sep="\t")
```


```{r}
summarised_gene_level_motif_and_expression
```

```{r}
unique_motif_id<-summarised_gene_level_motif_and_expression%>%select(motif_id)%>%extract2(2)%>%unique
```

```{r}
write.table(unique_motif_id, file="unique_motif_id_list.txt",sep="/n", row.names = F, quote=F)
```

##MOTIF EVALUATION
join motif from differnt cluster
```{bash}
#!/bin/sh
echo begin
date
#choose directory
dir_list=$(ls -d cluster_19/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd
echo merge upstream_motifs
meme_up=$(find $home_dir/$dir -path '*up/meme.html'| sort)
echo $meme_list
/usr/local/meme5/libexec/meme-5.0.0//meme2meme $meme_up -numbers -bg $home_dir/bg_up-model-1 > meme_up.txt
echo merge downstream_motifs
meme_down=$(find $home_dir/$dir -path '*down/meme.html'| sort)
echo $meme_list
/usr/local/meme5/libexec/meme-5.0.0//meme2meme $meme_down -numbers -bg $home_dir/bg_down-model-1 > meme_down.txt
done

```
find similar motif (self comparison and database comparison)

```{r}
#!/bin/sh
echo begin
date
#choose directory
dir_list=$(ls -d cluster_19/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd
echo merge upstream_motifs
meme_up=$(find $home_dir/$dir -path '*up/meme.html'| sort)
echo $meme_list
/usr/local/meme5/libexec/meme-5.0.0//meme2meme $meme_up -numbers -bg $home_dir/bg_up-model-1 > meme_up.txt
echo merge downstream_motifs
meme_down=$(find $home_dir/$dir -path '*down/meme.html'| sort)
echo $meme_list
/usr/local/meme5/libexec/meme-5.0.0//meme2meme $meme_down -numbers -bg $home_dir/bg_down-model-1 > meme_down.txt
done
pkncsk@bifx-cli:~/cluster_0720_WardD2$ cat tomtom.sh
#!/bin/sh
echo begin
date
#choose directory
dir_list=$(ls -d cluster_19/)
echo $dir_list
#loop for a folder of each clustering
for dir in $dir_list
do
home_dir=$(pwd)
cd "$home_dir/$dir"
echo current directory:
pwd
mkdir $home_dir/$dir/tomtom_out
echo self_comparison
mkdir $home_dir/$dir/tomtom_out/down_down/
/usr/local/meme5/bin/tomtom meme_down.txt meme_down.txt -oc $home_dir/$dir/tomtom_out/down_down/  -thresh 0.05 #-evalue  # -bfili $home_dir bg_down-model-1
mkdir $home_dir/$dir/tomtom_out/up_up/
/usr/local/meme5/bin/tomtom meme_up.txt meme_up.txt -oc $home_dir/$dir/tomtom_out/up_up/  -thresh 0.05 #-evalue # -bfile $home_dir bg_up-model-1
echo pair_comprison
mkdir $home_dir/$dir/tomtom_out/up_down/
/usr/local/meme5/bin/tomtom meme_up.txt meme_down.txt -oc $home_dir/$dir/tomtom_out/up_down/  -thresh 0.05 #-evalue # -bfile $home_dir bg_up-model-1
mkdir $home_dir/$dir/tomtom_out/down_up/
/usr/local/meme5/bin/tomtom meme_down.txt meme_up.txt -oc $home_dir/$dir/tomtom_out/down_up/ -thresh 0.05 #-evalue # -bfile $home_dir bg_down-model-1

echo jaspar-search
mkdir $home_dir/$dir/tomtom_out/down_JASPAR/
mkdir $home_dir/$dir/tomtom_out/up_JASPAR/
/usr/local/meme5/bin/tomtom meme_down.txt $home_dir/JASPAR2018_CORE_fungi_non-redundant_pfms_meme.txt -oc $home_dir/$dir/tomtom_out/down_JASPAR/ -thresh 0.05 #-evalue #-bfile $home_dir bg_down-model-1
/usr/local/meme5/bin/tomtom meme_up.txt $home_dir/JASPAR2018_CORE_fungi_non-redundant_pfms_meme.txt -oc $home_dir/$dir/tomtom_out/up_JASPAR/ -thresh 0.05 #-evalue #-bfile $home_dir bg_up-model-1
done

```

simplify motif table-collasp redundant motif into one group
using excel

```{r}
motif_cluster_data<-read.table("motif_group_non_redundant _motif_non_redundant.txt",sep ="\t",header=T)
```
```{r}
head(motif_cluster_data)
head(summarised_gene_level_motif_and_expression
)
```

join table to get the group info


```{r}
summarise_gene_level_with_motif_data<-summarised_gene_level_motif_and_expression%>%left_join(motif_cluster_data, by=c("motif_id"="ID"))
```

```{r}
colnames(summarise_gene_level_with_motif_data)
```


```{r}
summarise_gene_level_with_motif_data_work_table<-summarise_gene_level_with_motif_data[,c(1,150,3,9:12)]%>%group_by(gene, final_group)%>%summarise(accumulated_count=sum(count))
```
```{r}
summarise_gene_level_with_motif_data_work_table
```
join with expression table
```{r}
summarise_gene_level_with_motif_data_work_table_and_expression<-summarise_gene_level_with_motif_data_work_table%>% left_join(expression_table_with_cluster_number,by=c("gene"="rowname"))%>%select(-value)
```

```{r}
summarise_gene_level_with_motif_data_work_table_and_expression<-summarise_gene_level_with_motif_data_work_table_and_expression[,c(1:3,8:11)]
```

```{r}
summarise_gene_level_with_motif_data_work_table_and_expression
```
